### PR TITLE
instructions to install a syncer on a pcluster

### LINKED ISSUE #418: Prototype 3-4: PCluster Registration/Admin Demo
### Demo Objective
- [x] Document a repeatable process for registering a physical cluster with kcp
- [x] (stretch goal) Add support in the `kubectl kcp` plugin to register a physical cluster with kcp

### Demo Steps
1. User (or kcp admin, or org admin) needs to add new cluster (capacity, location, whatever)
    1. Choose/Create workspace (this will be the one the syncer will sync).
    2. Create SA in that workspace. The controller automatically generates a token in a `"token"` secret next to the SA.
    3. pcluster admin create kubeconfig with the upper token, and creates that as a secret in the pcluster syncer namespace
1. Reviews the yaml that creates the agent on the cluster
    1.  Sees reasonable roles+rolebindings, a deployment, and has to adapt the kubeconfig secret mount to match secret created above.
    2. YAML contains address of KCP to connect to
    3. YAML contains the unique name of this cluster (according to admin user)
    4. YAML references SA+token to authenticate to KCP, copied from KCP previously
1. Deploys the yaml 
    1. `kubectl apply -f syncer.yaml` 
    2. **Out-of-scope for p3** Possibly more automation around this
    3. Syncer starts by creating a `ClusterWorkspace` CR in KCP
1. Sees the WorkloadCluster show up in the workspace
1. **Out-of-scope for p3**: Does administrative action that makes this available to a workspace
1. Options:
    1. Decommission old cluster
        1. Triggers moves and drains
    1. Add new capacity
        1. The moment the WorkloadCluster is available, the new workload which wasn’t scheduled suddenly gets scheduled and placed (might be a stretch)
    1. Failure
        1. Sees an alert because the cluster is not responding? Don’t have any of our monitoring or conditions done (might be a stretch)

### Action Items

- [x] Scope the current demo as necessary to fit in prototype boundaries
- Technical tasks
   - [x] https://github.com/kcp-dev/kcp/pull/681 wire TokensController into pkg/server/server.go  @sttts  @marun 
   - ~~[ ] create WorkloadCluster object from syncer~~ find and document a procedure to make the syncer work through the service-account @aojea / @marun  
   - [x] build syncer controller that heartbeats @marun
   - [x] #738 build cluster-health controller in the server @imjasonh
   - [x] #723 @aojea 
   - [x] minimal, experimental/temporary RBAC policy on the pcluster
   - ~~[ ] **stretch-goal:** privileged syncer-namespace-controller and non-admin pcluster permissions for syncer~~https://github.com/kcp-dev/kcp/issues/387
   - [x] Out-of-scope (prototype 4 or 5): kcp plugin
   - Out-of-scope (prototype 4): any kind of in-cluster object with the resource list to sync
   - Out-of-scope (prototype 4): multi-workspace support (no "*" informers of any kind)
   - Out-of-scope: node-authorizer-like authorizer in kcp (probably replaced by virtual workspace)
   - Out-of-scope: lease management to avoid two syncer with the same cluster name
   - Out-of-scope: anything about bootstrap tokens, CSR, ...
- [x] ~~Contribute to final demo script and recording~~ part of @ncdc's youtube video https://www.youtube.com/watch?v=0trUxoQqdNk&t=1s
- ~~[ ] Remove KubeConfig from Cluster.Spec (TODO file issue for this)~~ https://github.com/kcp-dev/kcp/issues/934

### PR DESCRIPTION
**Pull request by aojea**
Part of #418

### CODE REVIEW

**Comment by ncdc** [comment]
FYI I removed `Fixes #418` from the description because that is an epic and there are more things needed before we can close it

File: contrib/syncer/README.md [hunk 1]
```diff
@@ -0,0 +1,31 @@
...
+
+We can see it has been created correctly and is ready:
+
+```sh
+kubectl get workspaces my-workspace -o wide
+NAME           TYPE        PHASE
+my-workspace   Universal   
+```
+
+2. ** FIXME ** Create a service account on the workspace
```
Comment (marun) [inline contrib/syncer/README.md:25]: Is it necessary to create a new service account in the namespace vs just using the default one that comes with a namespace?

File: contrib/syncer/syncer.yaml [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+    matchLabels:
+      app: kcp-syncer-ID
+  template:
+    metadata:
+      labels:
+        app: kcp-syncer-ID
+    spec:
+      containers:
+      - args:
+        - --from-kubeconfig=/kcp/kubeconfig
```
Comment (marun) [inline contrib/syncer/syncer.yaml:99]: I'm assuming this will be supplied as a secret and volume mounted?

File: contrib/syncer/syncer.yaml [hunk 3]
```diff
@@ -0,0 +1,125 @@
...
+        - --workload-cluster-name=WORKLOAD_CLUSTER_NAME   # ID of the -to cluster. Resources with this ID set in the 'workloads.kcp.dev/cluster' label will be synced
+        - --from-cluster=LOGICAL_CLUSTER                  # Name of the -from logical cluster
```
Comment (marun) [inline contrib/syncer/syncer.yaml:101]: Maybe note that this value can be found in the ClusterName field of the target workspace?

File: contrib/syncer/README.md [hunk 4]
```diff
@@ -0,0 +1,31 @@
+# KCP: Install Syncer component on the P-cluster
```
Comment (ncdc) [inline contrib/syncer/README.md:1]:
```suggestion
# Install syncer on a workload cluster
```

File: contrib/syncer/README.md [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+
+
+1. Create workspace on your Kcp cluster (remember to use your Kcp credentials)
```
Comment (ncdc) [inline contrib/syncer/README.md:4]:
```suggestion
1. Create a workspace
```

File: contrib/syncer/README.md [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+
+```sh
+cat <<EOF | kubectl --kubeconfig .kcp/admin.kubeconfig apply -f -
+apiVersion: tenancy.kcp.dev/v1beta1
+kind: Workspace
+metadata:
+  name: my-workspace
+spec:
+  type: Universal
+EOF
```
Comment (ncdc) [inline contrib/syncer/README.md:14]: Let's make our kcp kubectl plugin a prerequisite. Then tell users to do `kubectl kcp workspace create <name> --use`

File: contrib/syncer/README.md (hunk 1 above)
Comment (ncdc) [reply contrib/syncer/README.md:25]: It's probably better to have a new, dedicated SA just for the syncer, isn't it?

File: contrib/syncer/README.md [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+kubectl get workspaces my-workspace -o wide
+NAME           TYPE        PHASE
+my-workspace   Universal   
+```
+
+2. ** FIXME ** Create a service account on the workspace 
+
+```sh
+kubectl create namespace my-namespace
+kubectl create serviceaccount my-user --namespace my-namespace
```
Comment (ncdc) [inline contrib/syncer/README.md:29]:
```suggestion
kubectl create serviceaccount kcp-syncer --namespace my-namespace
```

File: contrib/syncer/syncer.yaml [hunk 8]
```diff
@@ -0,0 +1,125 @@
...
+  resources:
+  - namespaces
+  verbs:
+  - create
+- apiGroups:
+  - ""
+  # TODO set only the resources needed for the syncer
+  resources:
+  - services
+  - deployments
```
Comment (ncdc) [inline contrib/syncer/syncer.yaml:29]: Let's add `ingresses` too please

File: contrib/syncer/syncer.yaml [hunk 9]
```diff
@@ -0,0 +1,125 @@
...
+  verbs:
+  - get
+  - list
+  - watch
+  - create
+  - update
+  - delete
```
Comment (ncdc) [inline contrib/syncer/syncer.yaml:36]: Would `*` be any better?

File: contrib/syncer/syncer.yaml [hunk 10]
```diff
@@ -0,0 +1,125 @@
...
+roleRef:
+  apiGroup: rbac.authorization.k8s.io
+  kind: ClusterRole
+  name: syncer
+subjects:
+- kind: ServiceAccount
+  name: syncer
+  namespace: kcp-system
+---
+# Secret with the kubeconfig used to connect to the Kcp server
```
Comment (ncdc) [inline contrib/syncer/syncer.yaml:51]:
```suggestion
# Secret with the kubeconfig used to connect to the kcp server
```

File: contrib/syncer/syncer.yaml (hunk 8 above)
Comment (qiujian16) [inline contrib/syncer/syncer.yaml:29]: deployment/status also?

File: manifest/syncer.yaml [hunk 11]
```diff
@@ -0,0 +1,124 @@
...
+  namespace: kcp-system
+data:
+  kubeconfig: |-
+    apiVersion: v1
+    kind: Config
+    clusters:
+    - name: default-cluster
+      cluster:
+        certificate-authority-data: KCP_SYNCER_CACRT
+        server: KCP_SERVER_URL
```
Comment (sttts) [inline manifest/syncer.yaml:64]: use full URL

File: contrib/syncer/README.md (hunk 1 above)
Comment (sttts) [reply contrib/syncer/README.md]: +1

**Comment by aojea** [comment]
This is mostly done, blocked now in the following error

```
kubectl -n kcp-system logs syncer-77f85dfd7c-shnln
I0316 15:11:29.836320       1 syncer.go:66] Syncing the following resource types: [deployments.apps ingresses.networking.k8s services]
I0316 15:11:29.837723       1 specsyncer.go:73] Creating spec syncer for clusterName default:my-workspace--1 to pcluster WORKLOAD_CLUSTER_NAME, resources [deployments.apps ingresses.networking.k8s services]
E0316 15:11:29.841411       1 run.go:74] "command failed" err="Get \"https://192.168.1.150:6443/clusters/default:my-workspace--1/api?timeout=32s\": x509: certificate signed by unknown authority"
```

**Comment by aojea** [comment]
> Verify return code: 19 (self signed certificate in certificate chain)
:thinking:

**Comment by aojea** [comment]
cool, communication is working, blocked now in 

```
kubectl -n kcp-system logs syncer-7f854c985c-dl8np
I0316 16:07:10.912009       1 syncer.go:66] Syncing the following resource types: [deployments.apps]
I0316 16:07:10.912689       1 specsyncer.go:73] Creating spec syncer for clusterName default:my-workspace--1 to pcluster WORKLOAD_CLUSTER_NAME, resources [deployments.apps]
E0316 16:07:10.929702       1 run.go:74] "command failed" err="The following resource types were requested to be synced, but were not found in the KCP logical cluster: [deployments.apps]"
```

File: docs/kubectl_kcp.md [hunk 12]
```diff
@@ -0,0 +1,46 @@
+# Kubectl kcp plugin
```
Comment (sttts) [inline docs/kubectl_kcp.md:1]: nit: filename kubectl-kcp-plugin.md

File: docs/syncer.md [hunk 13]
```diff
@@ -0,0 +1,87 @@
...
+          default                                   root:default                              admin
+          root                                      root                                      admin
+          system:admin                              system:admin                              admin
+*         workspace.kcp.dev/personal/my-workspace   workspace.kcp.dev/personal/my-workspace   workspace.kcp.dev/personal/my-workspace
+```
+
+2. Create a service account for the syncer on current workspace:
+
+```sh
+kubectl create serviceaccount syncer
```
Comment (sttts) [inline docs/syncer.md:53]: `$`

File: docs/syncer.md [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+
+Obtain the KCP server URL and the logical cluster name:
+
+- KCP_SERVER_URL=https://192.168.1.150:6443
+- KCP_LOGICAL_CLUSTER_NAME=default:my-workspace--1
+
+
+** NOTE **
+
+You can use kubectl context to switch between different namespaces:
```
[the remark this answers is not in the document]
Comment (aojea) [reply docs/syncer.md:38]: yep

File: manifest/syncer.yaml [hunk 15]
```diff
@@ -0,0 +1,138 @@
...
+    spec:
+      containers:
+      - args:
+        - --from-kubeconfig=/kcp/config/kubeconfig
+        - --workload-cluster-name=WORKLOAD_CLUSTER_NAME   # ID of the -to cluster. Resources with this ID set in the 'workloads.kcp.dev/cluster' label will be synced
+        - --from-cluster=KCP_LOGICAL_CLUSTER_NAME         # Name of the -from logical cluster
+        - --sync-resources=deployments.apps               # List of resources to be synchronized in kcp
+        - --sync-resources=ingresses.networking.k8s
+        - --sync-resources=services
+        image: ghcr.io/kcp-dev/kcp/syncer-c2e3073d5026a8f7f2c47a50c16bdbec:7005a96@sha256:9735cbe04f4a5c7137e1f544fa29520f1624304629371d6bd185fcfe15f3b7af
```
Comment (sttts) [inline manifest/syncer.yaml:113]: should we reference latest?

**Review by sttts — APPROVED** [review]

**Comment by sttts** [comment]
Approved. Good as a first step. We will iterate.

**Pull request merged**